put.js

```diff
@@ -44,57 +46,49 @@ function putStream (cache, key, opts) {
   opts = PutOpts(opts)
   let integrity
   let size
-  const contentStream = write
-    .stream(cache, opts)
+
+  let memoData
+  const pipeline = new Pipeline()
+  // first item in the pipeline is the memoizer, because we need
+  // that to end first and get the collected data.
+  if (opts.memoize) {
+    const memoizer = new PassThrough().on('collect', data => {
+      memoData = data
+    })
+    pipeline.push(memoizer)
+  }
+
+  // contentStream is a write-only, not a passthrough
+  // no data comes out of it.
+  const contentStream = write.stream(cache, opts)
     .on('integrity', (int) => {
       integrity = int
     })
     .on('size', (s) => {
       size = s
     })
-  let memoData
-  let memoTotal = 0
-  const stream = to(
-    (chunk, enc, cb) => {
-      contentStream.write(chunk, enc, () => {
-        if (opts.memoize) {
-          if (!memoData) {
-            memoData = []
+
+  pipeline.push(contentStream)
+
+  // last but not least, we write the index and emit hash and size,
+  // and memoize if we're doing that
+  pipeline.push(new Flush({
+    flush () {
+      return index
+        .insert(cache, key, integrity, opts.concat({ size }))
+        .then((entry) => {
+          if (opts.memoize && memoData) {
+            memo.put(cache, entry, memoData, opts)
           }
-          memoData.push(chunk)
-          memoTotal += chunk.length
-        }
-        cb()
-      })
-    },
-    (cb) => {
-      contentStream.end(() => {
-        index
-          .insert(cache, key, integrity, opts.concat({ size }))
-          .then((entry) => {
-            if (opts.memoize) {
-              memo.put(cache, entry, Buffer.concat(memoData, memoTotal), opts)
-            }
-            stream.emit('integrity', integrity)
-            cb()
-          })
-      })
-    }
-  )
-  let erred = false
-  stream.once('error', (err) => {
-    if (erred) {
-      return
-    }
-    erred = true
-    contentStream.emit('error', err)
-  })
-  contentStream.once('error', (err) => {
-    if (erred) {
-      return
+          if (integrity) {
+            pipeline.emit('integrity', integrity)
+          }
+          if (size) {
+            pipeline.emit('size', size)
+          }
```

Comment:
this seems different from the original implementation (wasn't checking for `integrity` before `emit` and wasn't emitting `size`) I wonder if that affects anything 🤔

Reply:
Good catch.  I don't think it would cause a problem, and we almost always emit size along with integrity.  Probably fine?  At least, consistency is good.

Reply:
Huh, that is weird.  We emit `size` almost everywhere else.  I like the consistency, it's probably fine.

Reply:
sounds good 👍